tiledb/common/evaluator/evaluator.h

```diff
@@ -0,0 +1,411 @@
+};
+
+/**
+ * Policy that enforces a memory budget for the cache.
+ * This is considered to be the production policy which should
+ * be used in all real use cases.
+ *
+ * The memory consumption is accounted using a user provided function
+ * which returns how much a value uses plus the overhead of a shared_ptr.
+ * In the near future, the evaluator will be extended with orthogonal
+ * memory allocation policies for the values so the accounting for
+ * shared_ptr overhead might be conditioned by these policies.
+ *
+ * The budget is enforced by evicting LRU entries from the cache.
+ *
+ * @tparam Key The type of key
+ * @tparam Value The type of value
+ */
+
+template <
+    class Key,
+    class Value,
+    class SizeFn,
+    std::enable_if_t<
+        std::is_nothrow_invocable_r_v<
+            size_t,
+            SizeFn,
+            std::remove_cv_t<std::remove_reference_t<Value>>>,
+        bool> = true>
+class MemoryBudgetedCache : public virtual CachePolicyBase<Key, Value> {
+ public:
+  using return_type = std::shared_ptr<Value>;
+
+  MemoryBudgetedCache(SizeFn size_fn, size_t budget)
+      : size_fn_(size_fn)
+      , memory_budget_{budget}
+      , memory_consumed_(0) {
+    if (budget <= overhead_) {
+      throw std::logic_error(
+          "The memory budget must be greater than the minimum "
+          "memory overhead of the cache: " +
+          std::to_string(overhead_) + " bytes.");
+    }
+  }
+  /**
+   * Enforces the cache policy by evicting entries if necessary.
+   * Mass eviction might happen if the memory budget is low and the new
+   * value consumes a lot of memory. The algorithm for now is linear-time
+   * which should work ok when the cached values are of similar size, but we
+   * might need to change it in the future if it turns out value sizes are
+   * drastically different.
+   *
+   * This function takes the value produced by the callback
+   * to account its memory usage against the budget.
+   *
+   * @param v The value produced by the callback
+   */
+  virtual void enforce_policy(const Value& v) override {
+    auto mem_usage = overhead_ + size_fn_(v);
+    while (memory_consumed_ + mem_usage > memory_budget_) {
+      if (CachePolicyBase<Key, Value>::entries_.empty()) {
+        throw std::logic_error(
+            "The memory consumed by this value exceeds the budget of the "
+            "cache.");
+      }
+
+      auto evicted_value = this->evict_lru();
+      memory_consumed_ -= (size_fn_(*evicted_value) + overhead_);
+    }
+
+    // acount for the new value added in the cache
+    memory_consumed_ += mem_usage;
+  }
+
+ protected:
+  /** User provided function for calculating the size of a value */
+  SizeFn size_fn_;
+
+  /** The maximum amount of bytes managed by this cache */
+  size_t memory_budget_;
+
+  /** The maximum amount of bytes currently consumed by this cache */
+  size_t memory_consumed_;
+
+  static constexpr size_t overhead_ = sizeof(return_type);
```

Comment:
This would be better declared before the constructor, where it's used. It also definitely needs documentation.